src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs

```diff
@@ -0,0 +1,260 @@
+        {
+            bool? IsComVisibleCore(TypeDefinition typeDefinition)
+            {
+                CustomAttributeHandle handle = GetComVisibleAttribute(metadataReader, typeDefinition.GetCustomAttributes());
+                if (handle.IsNil)
+                {
+                    return null;
+                }
+
+                CustomAttribute comVisibleAttribute = metadataReader.GetCustomAttribute(handle);
+                CustomAttributeValue<KnownTypes> data = comVisibleAttribute.DecodeValue(new TypeResolver());
+                return (bool)data.FixedArguments[0].Value;
+            }
+
+            if (!definition.GetDeclaringType().IsNil)
+            {
+                return IsComVisible(metadataReader, metadataReader.GetTypeDefinition(definition.GetDeclaringType()), assemblyComVisible) && (IsComVisibleCore(definition) ?? assemblyComVisible);
+            }
+
+            return IsComVisibleCore(definition) ?? assemblyComVisible;
+        }
+
+        private static CustomAttributeHandle GetComVisibleAttribute(MetadataReader reader, CustomAttributeHandleCollection customAttributes)
+        {
+            foreach (CustomAttributeHandle attr in customAttributes)
+            {
+                CustomAttribute attribute = reader.GetCustomAttribute(attr);
+                MemberReference attributeConstructor = reader.GetMemberReference((MemberReferenceHandle)attribute.Constructor);
+                TypeReference attributeType = reader.GetTypeReference((TypeReferenceHandle)attributeConstructor.Parent);
+                if (reader.GetString(attributeType.Namespace) == "System.Runtime.InteropServices" && reader.GetString(attributeType.Name) == "ComVisibleAttribute")
+                {
+                    return attr;
+                }
+            }
+            return new CustomAttributeHandle();
+        }
+
+        private static string GetTypeGuid(MetadataReader reader, TypeDefinition type)
+        {
+            foreach (CustomAttributeHandle attr in type.GetCustomAttributes())
+            {
+                CustomAttribute attribute = reader.GetCustomAttribute(attr);
+                MemberReference attributeConstructor = reader.GetMemberReference((MemberReferenceHandle)attribute.Constructor);
+                TypeReference attributeType = reader.GetTypeReference((TypeReferenceHandle)attributeConstructor.Parent);
+                if (reader.GetString(attributeType.Namespace) == "System.Runtime.InteropServices" && reader.GetString(attributeType.Name) == "GuidAttribute")
+                {
+                    CustomAttributeValue<KnownTypes> data = attribute.DecodeValue(new TypeResolver());
+                    return (string)data.FixedArguments[0].Value;
+                }
+            }
+            throw new BuildErrorException(Strings.ClsidMapExportedTypesRequireExplicitGuid, GetTypeName(reader, type));
+        }
+
+        private enum KnownTypes
```

Comment:
This should not be plural.

Reply:
I mean it should be `enum KnownType`

---
In reply to: [249180190](https://github.com/dotnet/sdk/pull/2814#discussion_r249180190) [](ancestors = 249180190)